internal/net/call/call.go

```diff
@@ -188,19 +187,29 @@ type serverState struct {
 
 // Serve starts listening for connections and requests on l. Handlers to handle
 // incoming requests are found in hmap.
+//
+// Serve always returns a non-nil error and closes l.
 func Serve(ctx context.Context, l net.Listener, hmap *HandlerMap, opts ServerOptions) error {
 	opts = opts.withDefaults()
 	ss := &serverState{opts: opts}
 	defer ss.stop()
 
-	for ctx.Err() == nil {
+	// Arrange to close the listener when the context is canceled.
+	go func() {
+		<-ctx.Done()
+		l.Close()
+	}()
+
+	for {
 		conn, err := l.Accept()
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
 		if err != nil {
```

Comment:
If `l.Accept` fails for non-context related reasons and we get here, do we also have to close `l`? And then if we do, is it safe to race against the goroutine above that might also close `l`? We could use a `sync.Once` to close `l` if we're paranoid.

Reply:
Ah yes, great question! I think we do need to close `l` if we exit from some other error. It looks like `http.Server.Serve` has both `defer l.Close()` and `closeListenersLocked()` to close this listener, and uses `onceCloseListener` to wrap the close with a `sync.Once`.

Another option here might be to create a child context, cancel it when we receive a different error, and allow the goroutine to close `l`. In that case I think we need to wait on the goroutine to exit before exiting from `Serve`. 

I'll explore both of those options.

Reply:
I went with sync.Once, and copied the onceCloseListener from `net/http`. That seems more clear than adding a second context and a channel to wait on goroutine exit.